libraries/account-resolution/src/state.rs

```diff
@@ -0,0 +1,447 @@
+/// let program_id = Pubkey::new_unique();
+/// let mut instruction = Instruction::new_with_bytes(program_id, &[], vec![]);
+/// ExtraAccountMetas::add_to_instruction::<MyInstruction>(&mut instruction, &buffer).unwrap();
+///
+/// // On-chain, you can add the additional accounts *and* account infos
+/// let mut cpi_instruction = Instruction::new_with_bytes(program_id, &[], vec![]);
+/// let mut cpi_account_infos = vec![]; // assume the other required account infos are already included
+/// let remaining_account_infos: &[AccountInfo<'_>] = &[]; // these are the account infos provided to the instruction that are *not* part of any other known interface
+/// ExtraAccountMetas::add_to_cpi_instruction::<MyInstruction>(
+///     &mut cpi_instruction,
+///     &mut cpi_account_infos,
+///     &buffer,
+///     &remaining_account_infos,
+/// );
+/// ```
+pub struct ExtraAccountMetas;
+impl ExtraAccountMetas {
+    /// Initialize pod slice data for the given instruction and account infos
+    /// into the given buffer
+    pub fn init<'a, T: TlvDiscriminator, M: 'a>(
+        data: &mut [u8],
+        account_infos: &'a [M],
+    ) -> Result<(), ProgramError>
+    where
+        PodAccountMeta: From<&'a M>,
+    {
+        let mut state = TlvStateMut::unpack(data).unwrap();
+        let tlv_size = PodSlice::<PodAccountMeta>::size_of(account_infos.len())?;
+        let bytes = state.alloc::<T>(tlv_size)?;
+        let mut extra_account_metas = PodSliceMut::init(bytes)?;
+        for account_info in account_infos {
+            extra_account_metas.push(PodAccountMeta::from(account_info))?;
+        }
+        Ok(())
+    }
+
+    /// Initialize a TLV entry for the given discriminator, populating the data
+    /// with the given account infos
+    pub fn init_with_account_infos<T: TlvDiscriminator>(
+        data: &mut [u8],
+        account_infos: &[AccountInfo<'_>],
+    ) -> Result<(), ProgramError> {
+        Self::init::<T, AccountInfo>(data, account_infos)
+    }
+
+    /// Initialize a TLV entry for the given discriminator, populating the data
+    /// with the given account metas
+    pub fn init_with_account_metas<T: TlvDiscriminator>(
+        data: &mut [u8],
+        account_metas: &[AccountMeta],
+    ) -> Result<(), ProgramError> {
+        Self::init::<T, AccountMeta>(data, account_metas)
+    }
+
+    /// Get the byte size required to hold `num_items` items
+    pub fn size_of(num_items: usize) -> Result<usize, ProgramError> {
+        Ok(TlvStateBorrowed::get_base_len()
+            .saturating_add(PodSlice::<PodAccountMeta>::size_of(num_items)?))
+    }
+
+    /// Add the additional account metas to an existing instruction
+    pub fn add_to_instruction<T: TlvDiscriminator>(
+        instruction: &mut Instruction,
+        data: &[u8],
+    ) -> Result<(), ProgramError> {
+        let state = TlvStateBorrowed::unpack(data)?;
+        let bytes = state.get_bytes::<T>()?;
```

Comment:
Clean, so clean

Reply:
In this case we can't do it automatically unfortunately because the `PodSlice` is unsized. I spent awhile trying to come up with a slick way to do zero-copy serde for unsized types, but the lifetimes are really tricky to manage. In the end, TLV v1 keeps things super simple, but if we come up with a way to make it work in the TLV library, I'm super down to integrate it!